app/models/user.rb

```diff
@@ -0,0 +1,174 @@
+class User < ApplicationRecord
+  has_many :products
+  
+  validates :username, uniqueness: true, presence: true
+  validates :email, uniqueness: true, presence: true, format: { with: URI::MailTo::EMAIL_REGEXP }
+  
+  def order_count
+    orders = self.find_orders
+    return orders.count
+  end
+  
+  def top_product
+    max = 0
+    max_product = nil
+    self.products.each do |product|
+      
+      total_sold = product.count_sold()
+      
+      if total_sold > max
+        max = total_sold
+        max_product = product
+      end
+    end
+    
+    return max_product 
+  end
+  
+  def find_orders
+    matching_orders = self.products.flat_map{ |prod| prod.order_items }.map{ |oi| oi.order }.uniq
+    return matching_orders
+  end
+  
+  def total_revenue_by_order(order_id)
+    order = Order.find(order_id)
+    total_cost = self.order_revenue(order)
+    return total_cost
+  end
+  
+  def order_revenue(order)
+    total_per_order = 0
+    order.order_items.each do |oi|
+      product = Product.find_by(id: oi.product_id)
+      if product.user_id == self.id
+        total_per_order += oi.subtotal
+      end
+    end
+    return total_per_order
+  end
+  
+  def total_revenue()
+    orders = self.find_orders
+    
+    total_cost = 0.0
+    orders.each do |order|
+      total_cost += self.order_revenue(order)
+    end
+    
+    return total_cost
+  end
+  
+  # I want the total revenue for a set of orders based on status
+  # I have find_orders_by_status, which will return a list of orders by status.
+  # Then I need to find the total revenue for each of those orders, for my merchant
+  def total_revenue_by_status(status)
+    all_orders = self.find_orders_by_status(status)
+    
+    total_revenue = 0.0
+    if all_orders
+      all_orders.each do |order|
+        total_revenue += total_revenue_by_order(order.id)
+      end
+    end
+    
+    return total_revenue
+  end
+  
+  def find_orders_by_status(status)
+    if status == :all
+      return self.find_orders
+    else
+      orders_by_status = self.sort_orders_by_status()
+      return orders_by_status[status]
+    end
+  end
+  
+  def get_order_status(order)
+    result = find_order_statuses()
+    return result[:status_for_orders][order.id]
+  end
+  
+  def sort_orders_by_status()
+    result = find_order_statuses()
+    return result[:orders_by_status]
+  end
+  
+  def find_order_statuses()
```

Comment:
Per Sabrina's request for focused feedback. These group of methods look good. I wonder if `find_order_statuses` could have been broken up into smaller methods to simplify a bit.